deliaioanac/assignment2/printAncestors.cpp

```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include <vector>
+
+// Structure for a tree node
+struct TreeNode {
+    int val;
+    TreeNode *left;
+    TreeNode *right;
+    TreeNode *parent;
+    TreeNode(int x) : val(x), left(nullptr), right(nullptr), parent(nullptr) {}
+    TreeNode(int x, TreeNode *p) : val(x), left(nullptr), right(nullptr), parent(p) {}
+    ~TreeNode() {
+        delete left;
+        delete right;
+    }
+};
+
+// Class for binary tree
+class BinaryTree {
+    private:
+        TreeNode *tree;
+        TreeNode *dfsFind(TreeNode *node, const int val);
+    public:
+        TreeNode* createTree(const std::vector<int> &values, int k = 0, TreeNode *parent = nullptr);
+        void destroyTree();
+        void printAncestors(const int val);
+};
+
+// Populates binary tree with the values from the vector
+// The current node has the value at position k and a pointer to the parent node
+TreeNode* BinaryTree::createTree(const std::vector<int> &values, int k, TreeNode *parent) {
+    // Check if there are any values to use for the nodes
+    if (values.empty() || k > values.size()) {
+        return nullptr;
+    }
+
+    // If this is true, then this is the first iteration of the function,
+    // which initializes the tree member of the BinaryTree class
+    if (k == 0) {
+        tree = createTree(values, 1);
+        return nullptr;
+    }
+    
+    // Initialize the current node if kth value
+    TreeNode *node = new TreeNode(values[k - 1], parent);
+    // Initialize child nodes with appropriate values
+    node->left = createTree(values, k * 2, node);
+    node->right = createTree(values, k * 2 + 1, node);
+
+    return node;
+}
+
+// Destry the binary tree recursively
+void BinaryTree::destroyTree() {
+    delete tree;
+}
+
+// Returns the tree node that holds this value
+TreeNode* BinaryTree::dfsFind(TreeNode *node, const int val) {
+    // Check if the node could not be found in this (sub)tree
+    if (!node) {
+        return nullptr;
+    }
+
+    // Check if the node was found
+    if (node->val == val) {
+        return node;
+    }
+
+    // If the node was found in the left subtree, return it
+    // Otherwise, return the result of the search in the right subtree
+    TreeNode *result = dfsFind(node->left, val);
+    return (result ? result : dfsFind(node->right, val));
+}
+
+// Prints all ancestors of the node with this value, bottom-up
+void BinaryTree::printAncestors(const int val) {
```

Comment:
[optional] Your solution could be slightly improved :)

Currently it's O(N+N) in the worst case (if the tree resembles a linked list with just all right or all left children), it can be improved to just O(N) in the worst case if you track the parent nodes you're visiting in `dfsFind`. As a bonus you wouldn't need to store pointers to the parent node.

Reply:
You are right! However, I think the O(n + n) complexity remains if the function still prints the parents, since it would still have to go through the vector of parents and print them. I modified it to that, but let me know if I should just return the list.

Reply:
Yeah, you're correct! 

If you separate the algorithm into 'store in vector' and 'print' it's still O(N+N), but I think it's still a slight improvement on the original code because you're occupying less memory with not storing parent pointers and you now have a more 're-usable' function that not just prints the nodes, but actually returns them.

It seems a bit nit-picky and in a real world situation you probably don't really care if it's O(N) or O(N+N) or whether you're storing one extra pointer per value, but in the interview it's definitely something worth pointing out :)

Reply:
Makes sense, thanks for explaining!